Comment on the 0.4 release and the flaws of the process (#25)
input-output-hk/ouroboros-consensus: docs/ReleaseProcess.md

[contributing-to-a-project]: https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#Commit-Guidelines
[chap]: https://github.com/input-output-hk/cardano-haskell-packages

> # The first time we had to break this release process

Reviewer: It's great we record this, however, we might need to create another piece of documentation for the people that just want to either cut a release or understand our release process. If you think this makes sense we could create a new issue for this.

On the first half of April 2023, we were asked to get a new Consensus release

> from current `master` in order for it to be integrated into the node. However

Reviewer: Maybe `ourodoros-network@master` or something like that, and adding a footnote that mentions that the code was living in that repo at that time?

the version on `master` could not be released because it was dependent on some
changes from network that were not yet released.

> In order to make this clear, I am adding a `+` at the end of the version number

Reviewer: I don't know if we should use `I` here. Maybe: 
```suggestion
In order to make this clear, we will add a `+` at the end of the version number
```

to indicate that it contains unreleased changes on top of the version mentioned
(which would be in CHaP). Note that in the cabal version, the `+` is not present
so for cabal `0.3.1.0+` and `0.3.1.0` are both `0.3.1.0`.

> The release engineer that was integrating the node at that time, had pinned a
> specific commit in a `source-repository-packages` clause in his integration
> branch but by luck, several circumstances took place:

Reviewer: ```suggestion
The release engineer that was integrating the node at that time had pinned a
specific commit in a `source-repository-packages` clause in his integration
branch and the following unfortunate combination of factors took place:
```

- the commit was in `0.3.1.0+`
- only `ouroboros-consensus-diffusion` depended on the network changes
- `ouroboros-consensus` with his changes was, as mentioned above, at `0.3.1.0+`
- `ouroboros-consensus-diffusion 0.3.1.0+` would not have been buildable at the
  current CHap versions
- but `ouroboros-consensus-diffusion 0.3.1.0` (the one from CHaP) declared a
  dependency on `ouroboros-consensus 0.3.1.0`
- and by chance, the changes in `ouroboros-consensus 0.3.1.0+` (which actually
  were "Breaking changes") didn't break the compilation of
  `ouroboros-consensus-diffusion 0.3.1.0`
- so he could pull `ouroboros-consensus 0.3.1.0+` from the
  `source-repository-packages` and `ouroboros-consensus-diffusion 0.3.1.0` from
  CHaP

So in that situation we were breaking the bundle in a way that our previous
rules couldn't handle.

> Problem 1: We were unable to make a release from `master`, so we had to make
> the release from some previous commit as a side branch.

> > Problem 2: Our bundles were too rigid for the development that happens.

Reviewer: I'm not sure what this sentence means. I don't know if it could be expanded somehow.

> Problem 2.1: Sharing a repository with Network means some changes from network

> > will transpire to our packages, and network might not want to make a release

Reviewer: ```suggestion
> will affect our packages, and network might not want to make a release
```
